<!DOCTYPE html>
<html lang="pt-BR">
<head>
    <meta charset="UTF-8">
    <meta http-equiv="X-UA-Compatible" content="IE=edge">
    <meta name="viewport" content="width=device-width, initial-scale=1.0">
    <link rel="shortcut icon" href="../static/assets/icon.svg" type="image/x-icon" />
    <link rel="stylesheet" type="text/css" href="{{ url_for('static', filename='css/main/main.css') }}" />
	<link rel="stylesheet" type="text/css" href="{{ url_for('static', filename='css/main/modal.css') }}" />
	<link rel="stylesheet" type="text/css" href="{{ url_for('static', filename='css/main/dark-theme.css') }}" />
    <title> AnnoTEP </title>

	<script src="https://unpkg.com/@phosphor-icons/web"></script>
</head>
<body>
	<div class="menuMobile">
		<i class="ph-bold ph-list"></i>
	</div>

	<aside class="aside">
		<ul id="menu">
			<div class="logoAnnotep home" id="Home">
					<img src="static/assets/icon.svg" alt="Logo do Annotep">
			</div>
			<li class="li open home" id="Home">
				<img class="home-icon" src="static/assets/home.svg">
				<spam class="txt-link">Home</spam>
			</li>
			<li class="li" id="Download">
				<img class="download-icon" src="static/assets/download.svg">
				<spam class="txt-link">Download</spam>
			</li>
			<li class="li" id="Genome">
				<img class="Genome-icon" src="static/assets/dna-icon-white.svg">
				<spam class="txt-link">Genomes</spam>
			</li>
			<li class="li" id="Help">
				<img class="help-icon" src="static/assets/QuestionCircle.svg">
				<spam class="txt-link"> Help </spam>
			</li>
			<li class="li" id="About">
				<img class="about-icon" src="static/assets/InfoCircle.svg">
				<spam class="txt-link"> About </spam>
			</li>
			<li class="li" id="Contact">
				<img class="contact-icon" src="static/assets/contact-us.svg">
				<spam class="txt-link"> Contact </spam>
			</li>
		</ul>
	</aside>

	<div class="cloudImageWrapper">
		<img id="cloudImage" class="replace" src="static/assets/CloudMoon.svg">
	</div>

	<main class="main sectionsPage">
		<header class="header">
			<section class="container-header">
				<img id="logoImage" class="logo" src="static/assets/Logo.svg" alt="Logo TEP"/>
			</section>
		</header>

		<section class="container-main">
			<article class="description-screen">
				<p> AnnoTEP is a platform for annotating transposable elements in plants. Using an input file in <span>fasta</span> format containing the plant genome on a chromosomal scale, it is possible to obtain:</p>
			</article>

			<section class="container-explicative">
				<div class="box-explicative">
					<img class="img-dna" src="static/assets/dna-icon.svg">
					<span class="be1"><b>Annotation</b> of all Class I and Class II elements. </span>
				</div>
				<div class="box-explicative">
					<img class="img-phylogeny" src="static/assets/phylogeny.svg">
					<span class="be2"> <b>Data visualization</b> in graph format and phylogenetic trees. </span>
				</div>
				<div class="box-explicative box-email">
					<img class="img-email" src="static/assets/email-icon-light.svg">
					<span class="be3">A link with the results is sent <br> directly to your email. </span>
				</div>
				<div class="box-explicative box-file-results">
					<img class="img-file-results" src="static/assets/file-results.svg">
					<span class="be4">  
						Example of complete annotation.
						<a href="{{ url_for('result_example')}}" class="click-here be3" target="_blank"> Click here </a>
					</span>	
				</div>
				
			</section>
			<div class="rectangle">
				<div class="rectangle-green"></div>
			</div>
		</section>

		<section class="container-data-input">
			<div class="title-input">
				<p>Data input</p>
			</div>
			<section class="container-email">
				<div class="note">
					<span> 
						The download is limited to <b>30 MB</b>. The maximum number of tasks that can be run simultaneously on the platform is  <span><b>2</b></span>. 
						If this limit is reached, your task will be placed on a waiting list. If you have a large genome or want to run the annotation quickly without 
						a queue limit, we recommend downloading the tool to your local machine by going to <b>"Download"</b> in the side menu. 
						For more information, acess <b>"Help > Web Server"</b>. 
					</span>
				</div>

				<div class="session-title">
					<p>Email Adress </p>
				</div>
					<input id="email" class="email" type="email" pattern=".com" placeholder="enter email" onfocus="clearPlaceholder(this)" required/> 
			</section>

			<section class="container-genome">
				<div class="session-title">
					<p>Genome data</p>
				</div>
				<div class="box-input">
					<input type="text" id="fileNameInput" class="input-text" readonly disabled>

					<label for="inputdata" class="file-input">
						<p class="name-btn">Browser</p> 
						  <input type="file" class="input-data" accept=".fasta" id="inputdata" onchange="updateFileName()">
					</label>
					<span>(.fasta)</span>
				</div>
			</section>
	
			<section class="container-action">
				<div class="session-title">
					<p>Type annotation</p>
				</div>
				<div class="box-checkboxs">
					<div class="container-checkbox">
						<input type="checkbox" id="sine" name="sine" class="checkbox" checked>
						<label for="sine">SINE annotation</label>
					</div>
					<div class="container-checkbox">
						<input type="checkbox" id="line" name="line" class="checkbox">
						<label for="line">LINE annotation</label>
					</div>
					<div class="container-checkbox">
						<input type="checkbox" id="complete" name="complete" class="checkbox">
						<label for="complete">Complete annotation (SINE, LINE, LTR, TIR, Helitron, ...) <label>
					</div>	  
				</div>	
			</section>
	
			<div class="container-submit">
				<button type="button" id="uploaddata" class="uploaddata" disabled> Submit </button>
			</div>

			<p class="num_work">Annotation tasks in progress: <span>{{ num_workers }}</span></p>

		</section>

		<section id="myModal" class="modal">
			<div class="modal-content">
			  	<span class="close">&times;</span>
			  	<div class="container-notification">
					<p class="session-title session-modal">
				  		Annotation and analysis process
					</p>
					<div class="box-information">
				  		<p>The annotation process has started</p>
				  		<p>A confirmation email has been sent to your inbox containing further instructions.</p>
					</div>
			  	</div>
			</div>
		</section>

		<footer class="footer">
			<div class="cite-us">
				If you have found AnnoTEP useful, please <a href=""> <b>cite us</b> </a>
			</div>
		</footer>
	</main>

	<main class="Download">
		<header class="header">
			<section class="container-header">
				<img id="logoImage" class="logo" src="../static/assets/Logo.svg" alt="Logo TEP"/>
			</section>
		</header>

		<section class="container-download">
			<p> To ensure an efficient installation, we want to inform you that you need to have <a class="docker" href="https://docs.docker.com/engine/install/" target="_blank">Docker</a>  installed on your local machine. </p>
			<p> The Annotep platform has two local versions, the user-interface version and the version that works only via the machine's command terminal.</p>

			<div class="title-input title-download">
				<p>Graphic Interface</p>
			</div>

			<p>It has a similar interface to the web platform, but the difference lies in the delivery of the results, here they are saved in a local directory. In addition, this version relies on the internet to report the start and end of the annotation, or errors, via email.</p>
			<b>Installation guide</b>
			<ol class="installation-guide">
				<li>Download the latest annotep image</li>
				<span class="code-docker">docker pull annotep/graphic-interface:v1</span class="code-docker">
				<li>Run the following command</li>
				<span class="code-docker command-docker">docker run -it -v {folder-results}:/root/TEs/www/results -dp 0.0.0.0:5000:5000 annotep/graphic-interface:v1</span>
				<li>Example</li>
				<span class="code-docker command-docker">docker run -it -v $HOME/results-annotep:/root/TEs/www/results -dp 0.0.0.0:5000:5000 annotep/graphic-interface:v1</span>
			</ol>

			<p>To access the platform, open a browser and type in the address: <span>127.0.0.1:5000</span> </p>
			<ul class="results-annotation">
				<li>For more information, visit the tool's repository:<a href="https://github.com/Marcos-Fernando/AnnoTEP?tab=readme-ov-file#installation-with-container" target="_blank"> Github.</a></li>
			</ul>

			<div class="title-input title-download">
				<p>Bash interface</p>
			</div>

			<p>The focus of this version is direct use in the machine's command terminal without the need for an interface. It does not use the internet to run. In this version, the user must pay attention to the end of the annotation.</p>
			<b>Installation guide</b>
			<ol class="installation-guide">
				<li>Download the latest annotep image</li>
				<span class="code-docker">docker pull annotep/bash-interface:v1</span>

				<li>  Use the <b>-h </b> parameter to display a user guide describing how to use the script </li>
				<span class="code-docker"> 
					docker run annotep/bash-interface:v1 python run_annotep.py -h
				</span>

				<li>Run the container using the command below as a guide. </li>
				<ul>
					<li><p>We recommend creating a folder to insert your genomic data in FASTA format. </p></li>
					<li><p>Make sure you provide the full path to the folder where you want to save the results {folder-results} , as well as the full path to the genomes folder {folder-genomes/genome.fasta} :</p></li>
				</ul>
				
				<span class="code-docker command-docker"> 
					docker run -it -v {folder-results}:/root/TEs/results -v $HOME/TEs:{folder-genomes} 
					annotep/bash-interface:v1 python run_annotep.py --file {folder-genomes/genome.fasta} 
					--type {type-annotation} --threads {optional} 
				</span>

				<li> Examples</li>
				<p> <b>Annotation SINE or LINE</b> </p>
				<span class="code-docker command-docker"> 
					docker run -it -v $HOME/results-annotep:/root/TEs/results -v $HOME/TEs:$HOME/TEs 
					annotep/bash-interface:v1 python run_annotep.py --file $HOME/TEs/AtChr4.fasta 
					--type 2 
				</span>
				<p> <b>Annotation Complete</b> </p>
				<span class="code-docker command-docker"> 
					docker run -it -v $HOME/results-annotep:/root/TEs/results -v $HOME/TEs:$HOME/TEs 
					annotep/bash-interface:v1 python run_annotep.py --file $HOME/TEs/AtChr4.fasta 
					--type 4 --threads 12 
				</span>
			</ol>
			<ul class="results-annotation">
				<li>For more information, visit the tool's repository:<a href="https://github.com/Marcos-Fernando/AnnoTEP?tab=readme-ov-file#bash-interface" target="_blank"> Github.</a></li>
			</ul>
		</section>

		<footer class="footer">
			<div class="cite-us">
				
			</div>
	</main> 	

	<main class="Genome">
		<header class="header">
			<section class="container-header">
				<img id="logoImage" class="logo" src="../static/assets/Logo.svg" alt="Logo TEP"/>
			</section>
		</header>
		<section class="container-data-genome">
			<div class="title-input">
				<p>Preprocessed Genomes</p>
			</div>
			<section class="container-menu-genome">
				<div class="box-genomes-plants">
					<a href="{{ url_for('A_thaliana')}}" target="_blank">
						<div class="name-plants">
							<span><i> Arabidopsis thaliana </i></span>
						</div>
						<img class="image-plants" src="../static/screenshot/Arabidopsis_thaliana.jpg"/>
					</a>
				</div>
				<div class="box-genomes-plants">
					<a>
						<div class="name-plants">
							<span><i> Coming soon </i></span>
						</div>
						<img class="image-plants cs" src="../static/assets/coming-soon.png"/>
					</a>
				</div>
			</section>
		</section>
		<footer class="footer">
			<div class="cite-us">
				
			</div>
		</footer>
	</main>

	<main class="Help">
		<header class="header">
			<section class="container-header">
				<img id="logoImage" class="logo" src="../static/assets/Logo.svg" alt="Logo TEP"/>
			</section>
		</header>
		<section class="documentation-annotep">
			<h3> AnnoTEP documentation - <span>Anno</span>tation <span>T</span>ransposable <span>E</span>lements for <span>P</span>lants </h3>
			
			<p> Welcome to the AnnoTEP documentation. </p>
			<p>The AnnoTEP framework was developed to facilitate the annotation of transposable elements in plant genomes and to provide essential support to the scientific community involved in research centered on these elements.</p>
			<p> The tool has three distinct models: </p>
			<div class="menu-help">
				<ul>
					<li><p><a href="#web-server">Web Server</a></p></li>
					<li><p><a href="#container-gui">Container Graphical Interface</a></p></li>
					<li><p><a href="#container-bash">Container Bash Interface</a></p></li>
				</ul>
			</div>

			<div id="web-server" class="section-models">
				<h3>Web Server</h3>
				<p> 
					This model was developed to demonstrate how AnnoTEP works and to allow users to make small annotations of transposable elements 
					completely remotely (online). It is ideal for researchers who want to try out the tool without the need for local installation, 
					facilitating quick access and accommodating essential features for annotating TEs in plant genomes. 
				</p>
				<p>
					Due to the limitations of the web format, file sizes are restricted to 30MB, prioritizing up to two annotations at a time. 
					Annotations of files of this size can take up to three days to complete. For annotations of larger files, we recommend downloading the tool. 
					You can download it from the platform, in the Download section, or from <a href="https://github.com/Marcos-Fernando/AnnoTEP" target="_blank"> Github.</a>
				</p>
				<p> In this version, to generate the annotation, simply enter your e-mail address, the genomic file and the type of annotation you want, and submit. 
					You will receive information about the start and end of the annotation via email, as well as a link to access the results. Remember that these 
					results are only stored in the database for 48 hours. The genomic data requested for annotation must be in FASTA format. 
				</p>

				<ul class="results-annotation">
					<li><p><a href="#results-annotation">Click here </a> to find out more about the type of annotation and result.</p></li>
				</ul>
			
				<div id="container-gui" class="section-models">
				<h3>Container Graphical Interface</h3>
				<p>For this version, your machine must have Internet access.</p>
				<p>This version can be installed via Docker, as shown in the download section. It has a graphical interface and functions similar to the web server model, with the difference that it works entirely on the user's machine via localhost.</p>
				<p>There is no limit to the size of the genome, and the annotation speed will depend on the capacity of your local machine or server. Here you can enter the number of threads you want to use for the complete annotation. Although it works on the local machine, this model integrates an e-mail system that will notify you when the annotation is finished. Also, avoid shutting down the machine during the process, as this can interrupt data analysis. </p>


				<ul class="results-annotation">
					<li><p>For the type of annotation and result, <a href="#results-annotation"> click here.</a></p></li>
				</ul>
			</div>
		
			<div id="container-bash" class="section-models">
				<h3>Container Bash Interface</h3>
				<p>This version works entirely via the command line and does not require Internet access. Like the version with a graphical interface, it has no genome size limit, and the annotation speed will depend on the capacity of the local machine or server.</p>
				<p>Your data is entered using predefined parameters. All you have to do is enter the path of the genome on your machine, the type of annotation, and depending on the type of annotation, you can enter the number of threads you want to use. All these steps are described in the download section. </p>
				<p>It can be a little more complex to use than the version with a graphical interface.</p>

				<ul class="results-annotation">
					<li><p><a href="#results-annotation">Click here </a> to find out more about the type of annotation and result.</p></li>
				</ul>
			</div>

			<div id="results-annotation" class="section-models">
				<h3>Annotation types and results</h3>
				<ul class="seleting-annotation">
					<li><p><b>Selecting SINE Annotation: </b> This option exclusively analyzes the SINE elements present in the plant genome, generating data in .fa format and superimposed images of the element. In containerized versions, the result of this annotation is stored in the SINE folder.</p></li>
					<li><p><b>Selecting LINE Annotation: </b> This option exclusively analyzes the LINE elements present in the plant genome, generating libraries containing LINE sequences in .fa and .gff3 formats. In containerized versions, the result of this annotation is stored in the LINE folder. This option can generate some false positives, making it necessary to carry out the complete annotation to obtain more accurate data.</p></li>
					<li><p><b>Selecting SINE and LINE Annotation (Together): </b> Brings the same data generated by their unit versions.</p></li>
					<li><p><b>Selecting Complete Annotation: </b> This option performs an exhaustive analysis of various types of transposable elements, including SINEs, LINEs, LTR, TIR and Helitrons. The SINEs and LINEs data generated here is more refined than in the individual versions. In addition, the result generated by this annotation provides data on the TRIM, LARD, TR_GAG, BARE-2, MITES elements, as well as the Gypsy and Copia families. The analysis also generates graphs and phylogenetic trees of the LTRs, offering a comprehensive and detailed view of the transposable elements present in the genome studied. In addition, this annotation step masks the genome, helping to identify and suppress repetitive regions, which facilitates subsequent analysis and annotation of the genome.</p></li>
				</ul>

				<p>In the web server version, the results are stored in a database and presented in summary form. In the containerized versions, the results are more complete and detailed.</p>
				<ul class="results-annotation">
					<li><p>For more information, visit the tool's repository:<a href="https://github.com/Marcos-Fernando/AnnoTEP" target="_blank"> Github.</a></p></li>
				</ul>
			</div>
			
		</section>
		
		<footer class="footer">
			<div class="cite-us">
				
			</div>
		</footer>
	</main>	

	<main class="About">
		<header class="header">
			<section class="container-header">
				<img id="logoImage" class="logo" src="../static/assets/Logo.svg" alt="Logo TEP"/>
			</section>
		</header>

		<section class="description-about">
			<p> The AnnoTEP platform was developed to expand the range of tools available for annotating transposable elements (TEs) in plant genomes. Our main goal is to simplify and optimize the work of researchers, regardless of their level of experience, by facilitating the TE annotation process and providing as much information as possible about the genome. </p>
		
			<div class="title-input title-about">
				<p>Contributors</p>
			</div>

			<section class="table-logos">
				<section>
					<div class="logo-info">
						<img class="capes-logo" src="../static/screenshot/capes.png" alt="CAPES logo">
						<span>Coordenação de Aperfeiçoamento de Pessoal de Nível Superior </span>
					</div>
					<div class="logo-info">
						<img class="fapesp-logo" src="../static/screenshot/fapesp.png" alt="FAPESP logo">
						<span>Fundação de Amparo à Pesquisa do Estado de São Paulo</span>
					</div>	
	
					<div class="logo-info">
						<img class="fapespa-logo" src="../static/screenshot/fapespa.png" alt="FAPESPA logo">
						<span>Fundação Amazônia de Amparo a Estudos e Pesquisas</span>
					</div>	
				</section>

				<section>
					<div class="logo-info">
						<img class="ufpa-logo" src="../static/screenshot/ufpa.png" alt="UFPA logo">
						<span>Universidade Federal do Pará </span>
					</div>
	
					<div class="logo-info">
						<img class="unesp-logo" src="../static/screenshot/Logo_Unesp.svg" alt="UNESP logo">
						<span>Universidade Estadual Paulista</span>
					</div>
				</section>

				<section>
					<div class="logo-info">
						<img class="grid-logo" src="../static/screenshot/logo_gridunesp2.png" alt="GRID logo">
					</div>
	
					<div class="logo-info">
						<img class="ncc-logo" src="../static/screenshot/NCC.svg" alt="NCC logo">
						<span>Núcleo de Computação Científica Unesp</span>
					</div>
	
					<div class="logo-info">
						<img class="labiocad-logo" src="../static/screenshot/LaBioCAD.png" alt="LABIOCAD logo">
						<span>Laboratório de Bioinformatica e Computação de Alto desempenho</span>
					</div>
				</section>
			</section>
		</section>	

		<footer class="footer">
			<div class="cite-us">
				
			</div>
		</footer>
	</main>

	<main class="Contact">
		<header class="header">
			<section class="container-header">
				<img id="logoImage" class="logo" src="../static/assets/Logo.svg" alt="Logo TEP"/>
			</section>
		</header>

		<section class="description-contact">
			<div class="description-contact-text">
				<span>Contact us with any questions or to report a bug in the platform (web or local)</span>
			</div>

			<section class="container-email">
				<div class="session-title-contact">
					<p>Email Adress </p>
				</div>
					<input id="help-email" class="help-email" type="email" pattern=".com" placeholder="enter email" oninput="validateForm()" required/> 
			</section>

			<section class="container-email">
				<div class="session-title-contact">
					<p>Title</p>
				</div>
					<input id="help-title" class="help-email" type="email" placeholder="" oninput="validateForm()" required/> 
			</section>

			<section class="container-email">
				<div class="session-title-contact">
					<p>Subject</p>
				</div>
					<textarea id="help-subject" class="help-subject" type="text" placeholder="" oninput="validateForm()" required></textarea> 
			</section>

			<div class="container-submit">
				<button type="button" id="uploadcontact" class="uploaddata" onclick="submitForm()" disabled> Submit </button>
			</div>
		</section>

		<footer class="footer">
			<div class="cite-us">
				
			</div>
		</footer>
	</main>

	<script type="text/javascript" src="{{ url_for('static', filename='scripts/main/main.js') }}"></script>
	<script type="text/javascript" src="{{ url_for('static', filename='scripts/main/style.js') }}"></script>
	<script type="text/javascript" src="{{ url_for('static', filename='scripts/main/dark-theme.js') }}"></script>
	<script type="text/javascript" src="{{ url_for('static', filename='scripts/main/modal.js') }}"></script>
	<script type="text/javascript" src="{{ url_for('static', filename='scripts/main/pages.js') }}"></script>
</body>
</html>
